### PR TITLE
Drug Molecule Generation

### PR DESCRIPTION
**Pull request by victor369basu**
This is an example to demonstrate molecule generation with Variational Autoencoder.
Link - [Drug Molecule Generation](https://colab.research.google.com/drive/1u6uWJzs-aDVU7QP_7Xq4zEPB2ntcCj2p?usp=sharing)

For this example, we would be considered the research paper [Automatic chemical design using a data-driven continuous representation of molecules](https://arxiv.org/abs/1610.02415) as a reference. The model explained in this paper generate new molecules for efficient exploration and optimization through open-ended spaces of chemical compounds. The Model consists of three components: Encoder, Decoder and Predictor.  The encoder converts the discrete representation of a molecule into a real-valued continuous vector, and the decoder converts these continuous vectors back to discrete molecular representations. The predictor estimates chemical properties from the latent continuous vector representation of the molecule. Continuous representations allow the use of powerful gradient-based optimization to efficiently guide the search for optimized functional compounds.

### CODE REVIEW

File: examples/generative/molecule_generation.py [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+from rdkit import RDLogger
+from rdkit.Chem.Draw import IPythonConsole
+from rdkit.Chem.Draw import MolsToGridImage
+
+import tensorflow as tf
+from tensorflow.keras import layers
+
+import matplotlib.pyplot as plt
+
+tf.random.set_seed(2022)
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:82]: Please use `keras.utils.set_random_seed()` instead, it is more complete

File: examples/generative/molecule_generation.py [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+Virtual Screening**](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1360656/) dataset. The
+dataset comes with molecular formula in SMILE representation along with their respective
+molecular properties such as **logP**( Shows the water–octanal partition coefficient ),
+**SAS**( Shows the synthetic accessibility score ) and **QED**( Shows the Qualitative
+Estimate of Drug-likeness ).
+
+"""
+
+"""shell
+!wget
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:55]: There should be no line break here, and the `!` should not be included since it gets added automatically.

But if possible, try using `get_file()` in Python instead of relying on shell commands. Staying in Python is easier to manage.

File: examples/generative/molecule_generation.py [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+print(f"SMILES:\t{df.smiles[100]}\nlogP:\t{df.logP[100]}\nqed:\t{df.qed[100]}")
+molecule = molecule_from_smiles(df.iloc[100].smiles)
+print("Molecule:")
+molecule
+
+SMILE_CHARSET = '[" ", "#", ")", "(", "+", "-", "/", "1", "3", "2", "5", "4", "7", \
+                 "6", "=", "@", "C", "B", "F", "I", "H", "O", "N", "S", "[", "]", \
+                 "c", "l", "o", "n", "P", "s", "r"]'
+
+SMILE_CHARSET = ast.literal_eval(SMILE_CHARSET)
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:113]: Why is this necessary?

File: examples/generative/molecule_generation.py [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+test_df = df.drop(train_df.index)
+train_df.reset_index(drop=True, inplace=True)
+test_df.reset_index(drop=True, inplace=True)
+
+"""
+# Build the Encoder and Decoder
+"""
+
+
+class Encoder(tf.keras.Model):
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:222]: The encoder and decoder classes would be simpler and shorter if implemented as Functional models. Right now layers are defined/used twice. See: https://keras.io/examples/generative/vae/

You could package them as functions `get_encoder()` etc.

File: examples/generative/molecule_generation.py [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+The Categorical Crossentropy loss is to give a measure of the model's reconstruction
+capacity. The Property prediction loss is to estimate the mean squared error of the
+predicted and actual property after passing the ```z_mean(generated from the encoder)```
+through a property prediction model or property prediction layer.
+"""
+
+
+class MoleculeGenerator(tf.keras.Model):
+    def __init__(self, encoder, decoder, max_len, **kwargs):
+        super(MoleculeGenerator, self).__init__(**kwargs)
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:301]: Use Python 3 style super calls.

File: examples/generative/molecule_generation.py [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+
+import pandas as pd
+import numpy as np
+
+from rdkit import Chem
+from rdkit import RDLogger
+from rdkit.Chem.Draw import IPythonConsole
+from rdkit.Chem.Draw import MolsToGridImage
+
+import tensorflow as tf
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:77]: Import Keras too (`from tensorflow import keras`) and use it instead of `tf.keras`, that will save you some space

File: examples/generative/molecule_generation.py [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+            name="val_reconstruction_loss"
+        )
+        self.val_kl_loss_tracker = tf.keras.metrics.Mean(name="val_kl_loss")
+        self.val_pp_loss_tracker = tf.keras.metrics.Mean(
+            name="val_property_prediction_loss"
+        )
+
+    def compile(self, vae_optimizer, **kwargs):
+        super().compile(**kwargs)
+        self.vae_optimizer = vae_optimizer
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:327]: Why not just use `optimizer`? Why override `compile()`?

File: examples/generative/molecule_generation.py [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+                    tf.keras.losses.categorical_crossentropy(smile, reconstruction),
+                    axis=(1),
+                )
+            )
+            kl_loss = -0.5 * tf.reduce_sum(
+                1 + z_log_var - tf.square(z_mean) - tf.exp(z_log_var), 1
+            )
+            kl_loss = tf.reduce_mean(kl_loss)
+
+            pp_loss = tf.reduce_sum(
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:348]: Rather than using a small `pp` as variable name, use fully-spelled out variable names, for better readability

File: examples/generative/molecule_generation.py [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+
+vae_optimizer = tf.keras.optimizers.Adam(learning_rate=VAE_LR)
+
+encoder = Encoder(latent_dim)
+decoder = Decoder(latent_dim, max_length=MAX_LEN, charset_length=len(SMILE_CHARSET))
+
+model = MoleculeGenerator(encoder, decoder, MAX_MOLSIZE)
+
+model.compile(vae_optimizer)
+history = model.fit(train_loader, epochs=EPOCHS, validation_data=validation_loader)
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:436]: Does using `num_workers` improve performance when running on a multicore machine?

File: examples/generative/molecule_generation.py [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+plt.plot(history.history["loss"])
+plt.plot(history.history["val_loss"])
+plt.title("VAE Loss")
+plt.ylabel("Loss")
+plt.xlabel("epoch")
+plt.legend(["train", "val"], loc="upper left")
+plt.show()
+
+"""
+## Model Inferencing
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:447]: Just "Inference"

File: examples/generative/molecule_generation.py [hunk 11]
```diff
@@ -0,0 +1,493 @@
...
+"""
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:448]: Every section should have sufficient text explanations to make for an easy-to-follow (perhaps even compelling) read.

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR! This is a great topic. 

We have this related example already: https://keras.io/examples/generative/wgan-graphs/ -- we should reference it and highlight similarity and differences.

File: examples/generative/molecule_generation.py (hunk 3 above)
Comment (victor369basu) [reply examples/generative/molecule_generation.py]: To minimize the number of lines of code. I barely managed to find the entire code within 300 lines

File: examples/generative/molecule_generation.py (hunk 7 above)
Comment (victor369basu) [reply examples/generative/molecule_generation.py]: i have removed this snippet, I also felt its unnecessary

File: examples/generative/molecule_generation.py (hunk 9 above)
Comment (victor369basu) [reply examples/generative/molecule_generation.py]: I removed the data loader segment because it consumed too many lines of code. But I am not totally sure whether num_workers would improve model performance.

File: examples/generative/molecule_generation.py [hunk 12]
```diff
@@ -0,0 +1,575 @@
...
+                z_log_var, z_mean, qed_tensor, qed_pred, graph_real, graph_generated
+            )
+
+        grads = tape.gradient(total_loss, self.trainable_weights)
+        self.optimizer.apply_gradients(zip(grads, self.trainable_weights))
+
+        self.train_total_loss_tracker.update_state(total_loss)
+        return {"loss": self.train_total_loss_tracker.result()}
+
+    def calculate_loss(
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:423]: There is a built-in `compute_loss()` method, please use a slightly difference name to avoid confusion. `_calculate_loss()` maybe?

File: examples/generative/molecule_generation.py [hunk 13]
```diff
@@ -0,0 +1,575 @@
...
+
+    def call(self, inputs):
+        z_mean, log_var = self.encoder(inputs)
+        z = Sampling()([z_mean, log_var])
+
+        gen_adjacency, gen_features = self.decoder(z)
+
+        property_pred = self.property_prediction_layer(z_mean)
+
+        return z_mean, log_var, property_pred, gen_adjacency, gen_features
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:493]: Should we return a dict here? 5 different outputs is difficult to manage by position only.

File: examples/generative/molecule_generation.py [hunk 14]
```diff
@@ -0,0 +1,575 @@
...
+We would be inferring our model to predict over random latent space and try to generate
+100 new valid molecules.
+"""
+
+"""
+### Trying to generate 100 unique Molecules with the model.
+"""
+
+
+def sample(batch_size):
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:535]: You could make this a method on the model

File: examples/generative/molecule_generation.py [hunk 15]
```diff
@@ -0,0 +1,575 @@
...
+
+
+molecules = sample(100)
+
+MolsToGridImage(
+    [m for m in molecules if m is not None][:100], molsPerRow=5, subImgSize=(260, 160)
+)
+
+"""
+## Display the latent space clusters with respect to Molecular properties( QAE ).
```
Comment (fchollet) [inline examples/generative/molecule_generation.py:560]: Please add some words of commentary as conclusion

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

**Comment by victor369basu** [comment]
@fchollet I have updated the code as suggested.

File: examples/generative/molecule_generation.py [hunk 16]
```diff
@@ -0,0 +1,583 @@
...
+Alexander Kensert. Most of the functions referred to in this example are from the
+mentioned Keras example.
+"""
+
+"""
+## Setup
+"""
+
+"""shell
+!pip -q install rdkit-pypi==2021.9.4
```
Comment (fchollet) [inline examples/generative/molecule_generation.py]: Add comment to explain what this is for.

File: examples/generative/molecule_generation.py [hunk 17]
```diff
@@ -0,0 +1,583 @@
...
+* KL Loss
+* Property prediction Loss
+* Graph Loss( Gradient Penalty )
+
+The Categorical Crossentropy loss function is for giving a measure of the model's
+reconstruction capacity. The Property prediction loss is to estimate the mean squared
+error of the predicted and actual property after passing the `z_mean(generated from the
+encoder)` through a property prediction model or property prediction layer. The property
+prediction of the model is optimized through binary cross-entropy and the gradient
+penalty is further guided by the model's property(QED) prediction.
```
Comment (fchollet) [inline examples/generative/molecule_generation.py]: The last sentence is unclear, please rephrase. What is the gradient penalty for?

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the updates! I applied extensive copyedits -- please pull them first. 

You may want to improve the conclusion section.

Almost there :)

**Review by fchollet — COMMENTED** [review]
Thank you! I applied the final copyedits. Please add the generated files!

**Review by fchollet — APPROVED** [review]
Thanks for the great contribution!
